### PR TITLE
Refactor scheduler to enable switch on equivalence cache

### PR DESCRIPTION
**Pull request by resouer**
Part 2 of #30844 

Refactoring to enable easier switch on of equivalence cache.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34685)

### CODE REVIEW

**Comment by resouer** [comment]
ref https://github.com/kubernetes/kubernetes/issues/32024

**Comment by resouer** [comment]
PTAL @wojtek-t

File: plugin/pkg/scheduler/factory/factory.go [hunk 1]
```diff
@@ -78,6 +80,11 @@ type ConfigFactory struct {
 
 	scheduledPodPopulator *cache.Controller
 	nodePopulator         *cache.Controller
+	PodPopulator          *cache.Controller
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: This seems to be unused - please remove.

File: plugin/pkg/scheduler/factory/factory.go [hunk 2]
```diff
@@ -147,6 +157,84 @@ func NewConfigFactory(client *client.Client, schedulerName string, hardPodAffini
...
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(_ interface{}) {
+				c.EquivalencePodCache.SendClearAllCacheReq()
+			},
+			UpdateFunc: func(_, _ interface{}) {
+				c.EquivalencePodCache.SendClearAllCacheReq()
+			},
+			DeleteFunc: func(_ interface{}) {
+				c.EquivalencePodCache.SendClearAllCacheReq()
+			},
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Do we really need to clear the whole cache in these cases?

File: plugin/pkg/scheduler/factory/factory.go [hunk 3]
```diff
@@ -147,6 +157,84 @@ func NewConfigFactory(client *client.Client, schedulerName string, hardPodAffini
 		},
 	)
 
+	c.PVLister.Store, c.pvPopulator = cache.NewInformer(
+		c.createPersistentVolumeLW(),
+		&api.PersistentVolume{},
+		0,
+		cache.ResourceEventHandlerFuncs{
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: For each of those, please add a comment why we are clearning the cache (i.e. what predicate is invalidated).

File: plugin/pkg/scheduler/factory/factory.go (hunk 2 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Maybe we can some up with an API, to only clear the effected predicate from the cache in such cases.

File: plugin/pkg/scheduler/factory/factory.go [hunk 4]
```diff
@@ -172,6 +265,41 @@ func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
 		glog.Errorf("cannot convert newObj to *api.Pod: %v", newObj)
 		return
 	}
+
+	oldToleration, err := api.GetTolerationsFromPodAnnotations(oldPod.Annotations)
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: This whole logic should be encapsulated to a function.

File: plugin/pkg/scheduler/factory/factory.go [hunk 5]
```diff
@@ -182,13 +310,19 @@ func (c *ConfigFactory) deletePodFromCache(obj interface{}) {
...
+			c.EquivalencePodCache.SendInvalidAlgorithmCacheReq(pod.Spec.NodeName)
+		}
 	case cache.DeletedFinalStateUnknown:
 		var ok bool
 		pod, ok = t.Obj.(*api.Pod)
 		if !ok {
 			glog.Errorf("cannot convert to *api.Pod: %v", t.Obj)
 			return
 		}
+		if pod.Spec.NodeName != "" {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Instead of having this if everywhere, we should add a function to equivalenceCache, like: "invalidateCacheAfterPodChanges(*api.Pod)" (obviously with better name).

File: plugin/pkg/scheduler/factory/factory.go [hunk 6]
```diff
@@ -204,6 +338,10 @@ func (c *ConfigFactory) addNodeToCache(obj interface{}) {
 		glog.Errorf("cannot convert to *api.Node: %v", obj)
 		return
 	}
+
+	// have to invalid all equivalence cache when new node joined
+	c.EquivalencePodCache.SendClearAllCacheReq()
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Why? I don't understand it.

File: plugin/pkg/scheduler/factory/factory.go [hunk 7]
```diff
@@ -220,6 +358,11 @@ func (c *ConfigFactory) updateNodeInCache(oldObj, newObj interface{}) {
 		glog.Errorf("cannot convert newObj to *api.Node: %v", newObj)
 		return
 	}
+
+	// When node is updated, we need to invalid equivalence cache of all nodes.
+	// TODO(harryz) or only in some cases?
+	c.EquivalencePodCache.SendClearAllCacheReq()
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Why?

File: plugin/pkg/scheduler/factory/factory.go [hunk 8]
```diff
@@ -147,6 +157,84 @@ func NewConfigFactory(client *client.Client, schedulerName string, hardPodAffini
...
+					c.EquivalencePodCache.SendClearAllCacheReq()
+				}
+			},
+			DeleteFunc: func(_ interface{}) {
+				c.EquivalencePodCache.SendClearAllCacheReq()
+			},
+		},
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+	)
+
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: This PR is already quite huge, and is far from being trivial. Could you please split the more obvious parts to a separate PR.
I would like to see a PR only with:
- above changes to predicates.go
- adding these Populators here (but just with empty handler functions).

This PR will be definite not controversial and we would be able to merge that one fast, and continue working on this one.

Comment (resouer) [inline plugin/pkg/scheduler/factory/factory.go]: Yes, good idea, that makes scope much clear

**Comment by resouer** [comment]
Removed the cache update part, hopes we can review it with less effort.

File: plugin/pkg/scheduler/factory/factory.go [hunk 9]
```diff
@@ -147,15 +154,77 @@ func NewConfigFactory(client clientset.Interface, schedulerName string, hardPodA
 		},
 	)
 
+	// TODO(harryz) need to fill all the handlers here and below for equivalence cache
+	c.PVLister.Store, c.pvPopulator = cache.NewInformer(
+		c.createPersistentVolumeLW(),
+		&api.PersistentVolume{},
+		0,
+
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: nit: remove this empty line

File: plugin/pkg/scheduler/factory/factory.go [hunk 10]
```diff
@@ -147,15 +154,77 @@ func NewConfigFactory(client clientset.Interface, schedulerName string, hardPodA
...
+		cache.ResourceEventHandlerFuncs{
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: just do:
```
cache.ResourceEventHandlerFuncs{},
)
```

File: plugin/pkg/scheduler/factory/factory.go [hunk 11]
```diff
@@ -147,15 +154,77 @@ func NewConfigFactory(client clientset.Interface, schedulerName string, hardPodA
...
+			DeleteFunc: func(_ interface{}) {
+			},
+		},
+	)
+
+	c.PVCLister.Store, c.pvcPopulator = cache.NewInformer(
+		c.createPersistentVolumeClaimLW(),
+		&api.PersistentVolumeClaim{},
+		0,
+		cache.ResourceEventHandlerFuncs{
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: just do:
```
cache.ResourceEventHandlerFuncs{},
)
```

same below

File: plugin/pkg/scheduler/factory/factory.go [hunk 12]
```diff
@@ -320,9 +392,15 @@ func (f *ConfigFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
 		return nil, err
 	}
 
+	// Init equivalence class cache
+	if getEquivalencePodFunc != nil {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Can we move this change to the next PR?

File: plugin/pkg/scheduler/factory/factory.go [hunk 13]
```diff
@@ -320,9 +392,15 @@ func (f *ConfigFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
...
+		f.EquivalencePodCache = scheduler.NewEquivalenceCache(getEquivalencePodFunc)
+		glog.Info("Create equivalence class cache")
+	}
+
 	f.Run()
 
-	algo := scheduler.NewGenericScheduler(f.schedulerCache, predicateFuncs, priorityMetaProducer, priorityConfigs, extenders)
+	algo := scheduler.NewGenericScheduler(f.schedulerCache, f.EquivalencePodCache, predicateFuncs, priorityMetaProducer, priorityConfigs, extenders)
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Same for this one.

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 14]
```diff
@@ -376,16 +380,68 @@ func EqualPriorityMap(_ *api.Pod, _ interface{}, nodeInfo *schedulercache.NodeIn
...
+		equivalenceCache:     eCache,
 		predicates:           predicates,
 		priorityMetaProducer: priorityMetaProducer,
 		prioritizers:         prioritizers,
 		extenders:            extenders,
 		cachedNodeInfoMap:    make(map[string]*schedulercache.NodeInfo),
 	}
 }
+
+func (g *genericScheduler) scheduleWithEquivalenceClass(pod *api.Pod, nodes []*api.Node, trace *util.Trace) (string, error) {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/generic_scheduler.go]: Can we also move this to a separate PR?

**Comment by resouer** [comment]
@wojtek-t done, this PR is quite clean now

File: plugin/pkg/scheduler/factory/factory.go [hunk 15]
```diff
@@ -322,7 +365,7 @@ func (f *ConfigFactory) CreateFromKeys(predicateKeys, priorityKeys sets.String,
 
 	f.Run()
 
-	algo := scheduler.NewGenericScheduler(f.schedulerCache, predicateFuncs, priorityMetaProducer, priorityConfigs, extenders)
+	algo := scheduler.NewGenericScheduler(f.schedulerCache, nil, predicateFuncs, priorityMetaProducer, priorityConfigs, extenders)
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Can we move this change to a separate PR?

File: plugin/pkg/scheduler/factory/factory.go [hunk 16]
```diff
@@ -407,20 +450,15 @@ func (f *ConfigFactory) Run() {
...
-	cache.NewReflector(f.createPersistentVolumeClaimLW(), &api.PersistentVolumeClaim{}, f.PVCLister.Store, 0).RunUntil(f.StopEverything)
+	// Begin populating pv & pvc
+	go f.pvPopulator.Run(f.StopEverything)
+	go f.pvcPopulator.Run(f.StopEverything)
 
-	// Watch and cache all service objects. Scheduler needs to find all pods
-	// created by the same services or ReplicationControllers/ReplicaSets, so that it can spread them correctly.
-	// Cache this locally.
-	cache.NewReflector(f.createServiceLW(), &api.Service{}, f.ServiceLister.Indexer, 0).RunUntil(f.StopEverything)
+	// Begin populating pv & pvc
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Should be services.

**Comment by resouer** [comment]
@wojtek-t cleaned up

**Comment by wojtek-t** [comment]
lgtm  - thanks!

**Comment by resouer** [comment]
@wojtek-t I think we can discuss the cache updaters in #32024? Some pre-sudo code will be helpful before sending the next PR.
